libcaf_io/src/instance.cpp

```diff
@@ -110,156 +125,105 @@ connection_state instance::handle(execution_unit* ctx,
-        CAF_LOG_INFO("close connection since we already have a "
-                     "direct connection: " << CAF_ARG(hdr.source_node));
-        callee_.finalize_handshake(hdr.source_node, aid, sigs);
-        return close_connection;
-      }
-      // add direct route to this node and remove any indirect entry
-      CAF_LOG_INFO("new direct connection:" << CAF_ARG(hdr.source_node));
-      tbl_.add_direct(dm.handle, hdr.source_node);
-      auto was_indirect = tbl_.erase_indirect(hdr.source_node);
-      // write handshake as client in response
-      auto path = tbl_.lookup(hdr.source_node);
-      if (!path) {
-        CAF_LOG_ERROR("no route to host after server handshake");
-        return err();
-      }
-      write_client_handshake(ctx, path->wr_buf, hdr.source_node);
-      callee_.learned_new_node_directly(hdr.source_node, was_indirect);
-      callee_.finalize_handshake(hdr.source_node, aid, sigs);
-      flush(*path);
-      break;
+  // extract payload
+  std::vector<char> pl_buf{std::move_iterator<itr_t>(std::begin(dm.buf) +
+                                                     basp::header_size),
+                           std::move_iterator<itr_t>(std::end(dm.buf))};
+  // resize header
+  dm.buf.resize(basp::header_size);
+  // extract header
+  binary_deserializer bd{ctx, dm.buf};
+  auto e = bd(ep.hdr);
+  if (e || !valid(ep.hdr)) {
+    CAF_LOG_WARNING("received invalid header:" << CAF_ARG(ep.hdr));
+    //std::cerr << "[!!] invalid header!" << std::endl;
+    return err();
+  }
+  CAF_LOG_DEBUG(CAF_ARG(ep.hdr));
+  std::vector<char>* payload = nullptr;
+  if (ep.hdr.payload_len > 0) {
+    payload = &pl_buf;
+    if (payload->size() != ep.hdr.payload_len) {
+      CAF_LOG_WARNING("received invalid payload");
+      return err();
     }
-    case message_type::client_handshake: {
-      if (tbl_.lookup_direct(hdr.source_node) != invalid_connection_handle) {
-        CAF_LOG_INFO("received second client handshake:"
-                     << CAF_ARG(hdr.source_node));
-        break;
-      }
-      if (payload_valid()) {
-        binary_deserializer bd{ctx, *payload};
-        std::string remote_appid;
-        auto e = bd(remote_appid);
-        if (e)
-          return err();
-        if (remote_appid != callee_.system().config().middleman_app_identifier) {
-          CAF_LOG_ERROR("app identifier mismatch");
-          return err();
+  }
+  // Handle FIFO ordering of datagrams
+  if (ep.hdr.sequence_number > ep.seq_incoming) {
+    // Message arrived "early", add to pending messages
+    auto s = ep.hdr.sequence_number;
+    callee_.add_pending(s, ep, std::move(ep.hdr), std::move(pl_buf));
+    return true;
+  } else if (ep.hdr.sequence_number < ep.seq_incoming) {
+    // Message arrived late, drop it!
+    CAF_LOG_DEBUG("dropping msg " << CAF_ARG(dm));
+    return true;
+  }
+  // Message arrived as expected
+  ep.seq_incoming += 1;
+  // TODO: Add optional reliability here (send acks, ...)
```

Comment:
When this is merged we should create tickets for all these todos ;-)

Reply:
Probably. *[cough](https://github.com/actor-framework/actor-framework/search?utf8=%E2%9C%93&q=TODO&type=)*

Reply:
A couple of these todo's are from my ongoing streaming efforts and two are remainders to remove compiler workarounds when ready.

@Hamdor Here's a nice metaprogramming puzzle if you're up for a challenge: https://github.com/actor-framework/actor-framework/blob/85483265f9cb2908a31dbab38884867adaea1b28/libcaf_core/caf/send.hpp#L53 😉